File: packages/homebridge-plugin/lib/store.js
'use strict';

/**
 * DWM Store — Homebridge edition.
 *
 * Stores devices, DWM rules, and location in a single JSON file inside
 * Homebridge's storagePath (passed in at construction time, not via Electron).
 *
 * Schema mirrors the desktop store exactly so DWM rules created in the desktop
 * app can be imported / shared.
 *
 * ─── Data-loss protections (added in v2.0.36 after rules + devices vanished
 *     after a plugin upgrade) ───────────────────────────────────────────────
 *
 *  1. Distinguish "file does not exist" from "file exists but unreadable /
 *     unparseable".  The previous version returned DEFAULTS on every error,
 *     which caused the very next mutation to overwrite the file with empty
 *     state — destroying the user's data permanently.
 *
 *  2. Atomic writes via tmp + rename.  Eliminates the race window where a
 *     reader (the UI server process, or a heartbeat tick) catches the file
 *     mid-write and parses junk → returns DEFAULTS → overwrites.
 *
 *  3. Refuse to overwrite non-empty data with empty data.  If the in-memory
 *     state has zero devices AND zero rules AND no location, but the on-disk
 *     file already had real data, the write is rejected with a warning.
 *     Protects against process startup races between the plugin runtime and
 *     the UI server before they've both fully loaded.
 *
 *  4. Keep one rolling `.bak` of the last good file.  If the main file ever
 *     gets damaged, `dibby-wemo.json.bak` is one rename away.
 *
 *  5. Quarantine corrupt files instead of overwriting them.  A file that
 *     exists but fails to parse is moved to
 *     `dibby-wemo.json.corrupt-<unix-ts>` so the user can recover its
 *     contents manually.
 */

const path = require('path');
const fs   = require('fs');

const DEFAULTS = {
  location:           null,
  devices:            [],
  deviceGroups:       [],
  deviceOrder:        [],
  disabledRules:      {},
  dwmRules:           [],
  schedulerHeartbeat: null,
};

function isEmptyState(d) {
  if (!d) return true;
  const noDevices  = !Array.isArray(d.devices)  || d.devices.length  === 0;
  const noRules    = !Array.isArray(d.dwmRules) || d.dwmRules.length === 0;
  const noLocation = !d.location || (d.location.lat == null && d.location.lng == null);
  const noGroups   = !Array.isArray(d.deviceGroups) || d.deviceGroups.length === 0;
  const noOrder    = !Array.isArray(d.deviceOrder)  || d.deviceOrder.length  === 0;
  return noDevices && noRules && noLocation && noGroups && noOrder;
}

class DwmStore {
  constructor(storagePath) {
    this._filePath    = path.join(storagePath, 'dibby-wemo.json');
    this._bakPath     = `${this._filePath}.bak`;
    this._tmpPath     = `${this._filePath}.tmp`;
    // Cached snapshot of the last successful disk read.  Used both to short-
    // circuit hot paths (heartbeat saves) and as the "is the disk state
    // non-empty" signal for the empty-write guard.
    this._lastKnown   = null;
    this._safeToSave  = true;     // flipped false if we ever see a corrupt file
  }

  // ── Internal I/O ──────────────────────────────────────────────────────────

  /**
   * Read the file from disk.  Three outcomes:
   *  - Returns `{ ...DEFAULTS }` cleanly if the file simply doesn't exist
   *    (first-run case).
   *  - Returns parsed data if the file exists and is valid JSON.
   *  - If the file exists but is unreadable / corrupt, moves it aside to
   *    `*.corrupt-<ts>`, sets `_safeToSave = false`, and returns DEFAULTS.
   *    Subsequent `_save()` calls will be rejected until a clean reload
   *    succeeds.
   */
  _load() {
    let raw;
    try {
      raw = fs.readFileSync(this._filePath, 'utf8');
    } catch (e) {
      if (e.code === 'ENOENT') {
        // First-run: file simply doesn't exist yet.  Safe to save.
        this._safeToSave = true;
        return { ...DEFAULTS };
      }
      // EACCES, EBUSY, EIO, etc.  We don't know if the file holds user data,
      // so refuse to overwrite until we can read it cleanly.
      this._safeToSave = false;
      try {
        // eslint-disable-next-line no-console
        console.warn(`[DwmStore] cannot read ${this._filePath} (${e.code || e.message}). Refusing to save until readable.`);
      } catch { /* logging is best-effort */ }
      return { ...DEFAULTS };
    }

    let parsed;
    try {
      parsed = JSON.parse(raw);
    } catch {
      // File exists but isn't valid JSON.  Move it aside so a future write
      // doesn't destroy the original; user can inspect / recover it.
      const corruptPath = `${this._filePath}.corrupt-${Date.now()}`;
      try { fs.renameSync(this._filePath, corruptPath); } catch { /* best effort */ }
      try {
        // eslint-disable-next-line no-console
        console.warn(`[DwmStore] ${this._filePath} was unparseable JSON. Quarantined to ${corruptPath}. Will try .bak fallback.`);
      } catch { /* logging is best-effort */ }
      // Try the backup before declaring data lost.
      try {
        const bakRaw = fs.readFileSync(this._bakPath, 'utf8');
        parsed = JSON.parse(bakRaw);
        // Restore the .bak as the new main file.
        fs.writeFileSync(this._filePath, bakRaw, 'utf8');
        try { console.warn(`[DwmStore] recovered from ${this._bakPath}.`); } catch { /* */ }
      } catch {
        this._safeToSave = false;
        return { ...DEFAULTS };
      }
    }

    const merged = { ...DEFAULTS, ...parsed };
    this._lastKnown  = merged;
    this._safeToSave = true;
    return merged;
  }

  /**
   * Atomic save: write to a sibling .tmp file, then rename over the target.
   * Renames are atomic on POSIX filesystems and effectively-atomic on NTFS,
   * so readers never see a partially-written file.
   *
   * Refuses to write empty state on top of a non-empty file — that's the
   * single most common cause of "all my rules are gone" reports.
   */
  _save(data) {
    if (!this._safeToSave) {
      // eslint-disable-next-line no-console
      try { console.warn('[DwmStore] save skipped — last read was unsafe (corrupt or unreadable file).'); } catch { /* */ }
      return;
    }

    // Empty-write guard.  If we're about to flatten user data, abort and warn.
    if (isEmptyState(data) && this._lastKnown && !isEmptyState(this._lastKnown)) {
      // eslint-disable-next-line no-console
      try {
        console.warn(
          `[DwmStore] BLOCKED an empty-state write to ${this._filePath} — ` +
          `existing data (${(this._lastKnown.devices || []).length} device(s), ` +
          `${(this._lastKnown.dwmRules || []).length} DWM rule(s)) would be lost. ` +
          `Likely cause: process race between plugin runtime and UI server.`,
        );
      } catch { /* */ }
      return;
    }

    const payload = JSON.stringify(data, null, 2);
    try {
      // 1. Write to tmp.
      fs.writeFileSync(this._tmpPath, payload, 'utf8');
      // 2. Snapshot the previous file to .bak (best effort).
      try {
        if (fs.existsSync(this._filePath)) {
          fs.copyFileSync(this._filePath, this._bakPath);
        }
      } catch { /* .bak is best-effort */ }
      // 3. Atomic-rename tmp → target.
      fs.renameSync(this._tmpPath, this._filePath);
      this._lastKnown = data;
    } catch (e) {
      try { console.warn(`[DwmStore] save failed (${e.code || e.message}). Original ${this._filePath} left intact.`); } catch { /* */ }
      // Clean up the tmp file so it doesn't accumulate on disk.
      try { fs.unlinkSync(this._tmpPath); } catch { /* */ }
    }
  }

  // ── Location ──────────────────────────────────────────────────────────────

  getLocation()    { return this._load().location; }
  setLocation(loc) { const d = this._load(); d.location = loc; this._save(d); }

  // ── Devices ───────────────────────────────────────────────────────────────

  getDevices()              { return this._load().devices ?? []; }
  saveDevices(list)         { const d = this._load(); d.devices = list; this._save(d); }
  getDeviceOrder()          { return this._load().deviceOrder ?? []; }
  saveDeviceOrder(order)    { const d = this._load(); d.deviceOrder = order; this._save(d); }
  getDeviceGroups()         { return this._load().deviceGroups ?? []; }
  saveDeviceGroups(groups)  { const d = this._load(); d.deviceGroups = groups; this._save(d); }

  /**
   * Merge freshly discovered devices into the cached list.
   *
   * STICKY-DEVICE GUARANTEE — once a Wemo has been detected and stored, it
   * stays in the device list permanently.  A re-scan never removes anything:
   * if a device is offline / unreachable / on a different VLAN, its cached
   * record is preserved exactly as-is (host, port, UDN, friendlyName, model,
   * firmwareVersion).  Plugin upgrades (`npm update -g homebridge-dibby-wemo`)
   * leave this file untouched because it lives in Homebridge's storagePath,
   * outside the npm package directory.
   *
   * Behaviour:
   * - Existing devices get their host/port/name/firmware refreshed if a new
   *   scan returned newer values; everything else on the record is kept.
   * - Previously cached devices NOT in the new scan are kept verbatim
   *   (offline ≠ removed).
   * - Brand-new devices are appended.
   *
   * Returns the merged list (= the new on-disk state).
   */
  mergeDevices(fresh) {
    const d       = this._load();
    const cached  = d.devices ?? [];
    const byUdn   = new Map(cached.map((dev) => [dev.udn, dev]));

    for (const f of fresh) {
      const udn = f.udn ?? `${f.host}:${f.port}`;
      if (byUdn.has(udn)) {
        // Update existing entry with latest network data
        byUdn.set(udn, { ...byUdn.get(udn), ...f, udn });
      } else {
        byUdn.set(udn, { ...f, udn });
      }
    }

    d.devices = Array.from(byUdn.values());
    this._save(d);
    return d.devices;
  }

  // ── Disabled-rule backups ─────────────────────────────────────────────────

  getDisabledRules()                    { return this._load().disabledRules ?? {}; }
  setDisabledRule(key, ruleDevicesRows) {
    const d = this._load();
    if (!d.disabledRules) d.disabledRules = {};
    d.disabledRules[key] = ruleDevicesRows;
    this._save(d);
  }
  clearDisabledRule(key) {
    const d = this._load();
    if (!d.disabledRules) return;
    delete d.disabledRules[key];
    this._save(d);
  }

  // ── DWM Rules ─────────────────────────────────────────────────────────────

  getDwmRules() { return this._load().dwmRules ?? []; }

  createDwmRule(rule) {
    const d = this._load();
    if (!d.dwmRules) d.dwmRules = [];
    const id  = `dwm-${Date.now()}-${Math.random().toString(36).slice(2, 7)}`;
    const now = new Date().toISOString();
    const newRule = { ...rule, id, createdAt: now, updatedAt: now };
    d.dwmRules.push(newRule);
    this._save(d);
    return newRule;
  }

  updateDwmRule(id, updates) {
    const d = this._load();
    if (!d.dwmRules) d.dwmRules = [];
    const idx = d.dwmRules.findIndex((r) => r.id === id);
    if (idx === -1) throw new Error(`DWM rule not found: ${id}`);
    d.dwmRules[idx] = { ...d.dwmRules[idx], ...updates, id, updatedAt: new Date().toISOString() };
    this._save(d);
    return d.dwmRules[idx];
  }

  deleteDwmRule(id) {
    const d = this._load();
    if (!d.dwmRules) return;
    d.dwmRules = d.dwmRules.filter((r) => r.id !== id);
    this._save(d);
  }

  // ── Scheduler heartbeat ───────────────────────────────────────────────────

  getHeartbeat() { return this._load().schedulerHeartbeat ?? null; }

  saveHeartbeat(hb) {
    const d = this._load();
    d.schedulerHeartbeat = { ...hb, ts: new Date().toISOString() };
    this._save(d);
  }
}

module.exports = DwmStore;

File: custom_components/dibby_wemo/store.py
"""
JSON persistence store for Dibby Wemo.

Saves to ``<ha_config_dir>/dibby-wemo.json`` — schema-compatible with the
Homebridge plugin and desktop app stores.

All file I/O is dispatched to Home Assistant's executor so the event loop
is never blocked. The store is created via :meth:`DwmStore.async_create`
(an async factory) and every mutation schedules a fire-and-forget save on
the executor with a deep-copied snapshot of the in-memory data, so the
executor sees a consistent view even if subsequent mutations happen
before its write completes.
"""

from __future__ import annotations

import copy
import json
import logging
import os
import shutil
import time
from typing import Any

from homeassistant.core import HomeAssistant

from .const import STORE_FILENAME

_LOGGER = logging.getLogger(__name__)


class DwmStore:
    def __init__(self, hass: HomeAssistant, config_dir: str) -> None:
        self._hass = hass
        self._path = os.path.join(config_dir, STORE_FILENAME)
        self._data: dict[str, Any] = {}
        self._safe_to_save: bool = True
        self._last_known: dict[str, Any] | None = None

    @classmethod
    async def async_create(cls, hass: HomeAssistant, config_dir: str) -> "DwmStore":
        """Async factory — runs the initial blocking read on the executor."""
        self = cls(hass, config_dir)
        self._data = await hass.async_add_executor_job(self._sync_load)
        return self

    # ── Sync I/O primitives (run only inside executor) ───────────────────────

    # Data-loss protections — mirrors the JS DwmStore hardening:
    #  - distinguish missing file (safe) from unreadable/corrupt (never
    #    overwrite what we couldn't read)
    #  - atomic writes via tmp + os.replace (readers never see partial JSON)
    #  - rolling .bak of the last good file, used for recovery
    #  - refuse to flatten non-empty on-disk data with empty in-memory state

    @staticmethod
    def _is_empty_state(d: dict | None) -> bool:
        if not d:
            return True
        return not d.get("devices") and not d.get("dwmRules") and not d.get("location")

    def _sync_load(self) -> dict[str, Any]:
        try:
            with open(self._path, encoding="utf-8") as f:
                data = json.load(f)
            self._safe_to_save = True
            self._last_known = data
            return data
        except FileNotFoundError:
            self._safe_to_save = True
            return {}
        except json.JSONDecodeError:
            # Corrupt JSON — quarantine the original, try the .bak.
            corrupt = f"{self._path}.corrupt-{int(time.time())}"
            try:
                os.replace(self._path, corrupt)
                _LOGGER.warning("Corrupt store quarantined to %s", corrupt)
            except OSError:
                pass
            try:
                with open(f"{self._path}.bak", encoding="utf-8") as f:
                    data = json.load(f)
                with open(self._path, "w", encoding="utf-8") as f:
                    json.dump(data, f, indent=2)
                _LOGGER.warning("Store recovered from %s.bak", self._path)
                self._safe_to_save = True
                self._last_known = data
                return data
            except (OSError, json.JSONDecodeError):
                self._safe_to_save = False
                return {}
        except OSError as e:
            # EACCES / EBUSY / EIO — refuse to save over data we can't read.
            _LOGGER.warning("Cannot read %s (%s) — refusing to save until readable", self._path, e)
            self._safe_to_save = False
            return {}

    def _sync_save(self, snapshot: dict[str, Any]) -> None:
        if not getattr(self, "_safe_to_save", True):
            _LOGGER.warning("Save skipped — last read of %s was unsafe", self._path)
            return
        last = getattr(self, "_last_known", None)
        if self._is_empty_state(snapshot) and last and not self._is_empty_state(last):
            _LOGGER.warning(
                "BLOCKED empty-state write to %s — %d device(s), %d rule(s) would be lost",
                self._path, len(last.get("devices") or []), len(last.get("dwmRules") or []),
            )
            return
        tmp = f"{self._path}.tmp"
        try:
            with open(tmp, "w", encoding="utf-8") as f:
                json.dump(snapshot, f, indent=2)
            try:
                if os.path.exists(self._path):
                    shutil.copyfile(self._path, f"{self._path}.bak")
            except OSError:
                pass  # .bak is best-effort
            os.replace(tmp, self._path)  # atomic on POSIX and NTFS
            self._last_known = snapshot
        except OSError as e:
            _LOGGER.error("Could not save %s: %s — original left intact", self._path, e)
            try:
                os.unlink(tmp)
            except OSError:
                pass

    # ── Internal — dispatch async save ───────────────────────────────────────

    def _save(self) -> None:
        snapshot = copy.deepcopy(self._data)
        if self._hass is not None:
            self._hass.async_add_executor_job(self._sync_save, snapshot)
        else:  # pragma: no cover — defensive fallback for non-HA uses
            self._sync_save(snapshot)

    def _get(self, key: str, default=None):
        return self._data.get(key, default)

    def _set(self, key: str, value: Any) -> None:
        self._data[key] = value
        self._save()

    # ── Location ──────────────────────────────────────────────────────────────

    def get_location(self) -> dict | None:
        return self._get("location")

    def set_location(self, loc: dict) -> None:
        self._set("location", loc)

    # ── Devices ───────────────────────────────────────────────────────────────

    def get_devices(self) -> list[dict]:
        return self._get("devices", [])

    def save_devices(self, devices: list[dict]) -> None:
        self._set("devices", devices)

    def merge_devices(self, fresh: list[dict]) -> list[dict]:
        """Upsert fresh devices into saved list by UDN."""
        existing = {d["udn"]: d for d in self.get_devices() if d.get("udn")}
        for dev in fresh:
            udn = dev.get("udn")
            if udn:
                existing[udn] = {**existing.get(udn, {}), **dev}
        merged = list(existing.values())
        self.save_devices(merged)
        return merged

    # ── DWM Rules ─────────────────────────────────────────────────────────────

    def get_dwm_rules(self) -> list[dict]:
        return self._get("dwmRules", [])

    def create_dwm_rule(self, rule: dict) -> dict:
        rules = self.get_dwm_rules()
        now = _iso_now()
        uid = f"dwm-{int(time.time() * 1000)}-{os.urandom(3).hex()}"
        new_rule = {**rule, "id": uid, "createdAt": now, "updatedAt": now}
        rules.append(new_rule)
        self._set("dwmRules", rules)
        return new_rule

    def update_dwm_rule(self, rule_id: str, updates: dict) -> dict | None:
        rules = self.get_dwm_rules()
        for i, r in enumerate(rules):
            if r.get("id") == rule_id:
                rules[i] = {**r, **updates, "updatedAt": _iso_now()}
                self._set("dwmRules", rules)
                return rules[i]
        return None

    def delete_dwm_rule(self, rule_id: str) -> bool:
        rules = self.get_dwm_rules()
        new_rules = [r for r in rules if r.get("id") != rule_id]
        if len(new_rules) == len(rules):
            return False
        self._set("dwmRules", new_rules)
        return True

    # ── Heartbeat ─────────────────────────────────────────────────────────────

    def get_heartbeat(self) -> dict | None:
        return self._get("schedulerHeartbeat")

    def save_heartbeat(self, data: dict) -> None:
        self._set("schedulerHeartbeat", {**data, "ts": _iso_now()})


def _iso_now() -> str:
    from datetime import datetime
    return datetime.utcnow().isoformat() + "Z"
